Draft 3 Extended views for NDT7 using materialized joins and Geo renaming

mattmathis opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/etl-schema/91)
<!-- Reviewable:end -->

mattmathis commented on views/ndt/referenced-by/referenced-by/unified_downloads_20201026.sql:
```diff
@@ -0,0 +1,47 @@
+--
```
I need change this and the legacy uploads file names.  The BQ UI treats objects ending in underscore-datecode as single partitions of a partitioned table.

mattmathis commented on views/ndt/referenced-by/referenced-by/unified_uploads_20201026.sql:
```diff
@@ -0,0 +1,47 @@
+--
+-- This legacy view recreates the similarly named standard view as of
```
Bad file name, see above.

mattmathis commented:
*[a discussion](https://reviewable.io/reviews/m-lab/etl-schema/91#-MKYLGC66ke9Gdx9tcBd:-MKYLGC66ke9Gdx9tcBe:b-mtkubh) (no related file):*
I also created a draft email announcement, which might make this easier to understand.   See https://docs.google.com/document/d/1ZD4FkjSv35b7tuKj1voRZgfdQIA5_DHAm84xcvY3WCs/edit#

---


<!-- Sent from Reviewable.io -->

stephen-soltesz reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/m-lab/etl-schema/91)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @critzo, @gfr10598, @mattmathis, and @stephen-soltesz)

*[a discussion](https://reviewable.io/reviews/m-lab/etl-schema/91#-MKYLGC66ke9Gdx9tcBd:-MKe_vaAEp03aQwwBvqF:b-utzty7) (no related file):*
<details><summary><i>Previously, mattmathis (Matt Mathis) wrote…</i></summary><blockquote>

I also created a draft email announcement, which might make this easier to understand.   See https://docs.google.com/document/d/1ZD4FkjSv35b7tuKj1voRZgfdQIA5_DHAm84xcvY3WCs/edit#
</blockquote></details>

In the limit, I would like to see names like:

```
raw_ndt.ndt7                                                                    
raw_ndt.tcpinfo                                                                 
raw_ndt.traceroute                                                              
raw_ndt.annotation                                                              
                                                                                
intermediate_ndt.joined_ndt7                                                    
intermediate_ndt.joined_ndt5                                                    
intermediate_ndt.joined_web100                                                  
intermediate_ndt.extended_ndt7                                                  
intermediate_ndt.extended_ndt5                                                  
intermediate_ndt.extended_web100                                                
                                                                                
ndt.ndt7_uploads                                                                
ndt.ndt7_downloads                                                              
ndt.ndt5_uploads                                                                
ndt.ndt5_downloads                                                              
ndt.web100_uploads                                                              
ndt.web100_downloads                                                            
ndt.unified_uploads                                                             
ndt.unified_downloads
```

The split of upload/download doesn't need to happen until the last step. This still needs to remain manageable. This is getting more complex not less.

In hindsight, it's not clear that we want to publish the sidecar raw tables under `ndt.*` yet.

---
___
*[views/intermediate_ndt/referenced-by/extended_ndt5_downloads.sql, line 124 at r1](https://reviewable.io/reviews/m-lab/etl-schema/91#-MKe_Fg9EcxFUZXuwWdI:-MKe_Fg9EcxFUZXuwWdJ:bjbfr0k) ([raw file](https://github.com/m-lab/etl-schema/blob/23c70815e71c141d0a764181885e8b473f7f1606/views/intermediate_ndt/referenced-by/extended_ndt5_downloads.sql#L124)):*
> ```TSQL
>              client.Geo.radius, -- aka AccuracyRadiusKm
>              FALSE AS Missing -- Future missing record flag
>       ) AS new_Geo,  -- Future primary Geo
> ```

It is a mistake to introduce new record names here. The new fields should continue working without query changes after we retire the old fields. This is how the introduction of `date` and retirement of `test_date` worked.

For queries that currently use `Geo.continent_code`, we can either break now (and refer to the _<date> legacy table) or combine with the camel case so `Geo.ContinentCode` also works. We don't want users to have to change their queries twice if they adopt the new names now.


<!-- Sent from Reviewable.io -->

mattmathis commented:
*[a discussion](https://reviewable.io/reviews/m-lab/etl-schema/91#-MKYLGC66ke9Gdx9tcBd:-MKluxsH2NlbqouJ6w4j:b-dn97ai) (no related file):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

In the limit, I would like to see names like:

```
raw_ndt.ndt7                                                                    
raw_ndt.tcpinfo                                                                 
raw_ndt.traceroute                                                              
raw_ndt.annotation                                                              
                                                                                
intermediate_ndt.joined_ndt7                                                    
intermediate_ndt.joined_ndt5                                                    
intermediate_ndt.joined_web100                                                  
intermediate_ndt.extended_ndt7                                                  
intermediate_ndt.extended_ndt5                                                  
intermediate_ndt.extended_web100                                                
                                                                                
ndt.ndt7_uploads                                                                
ndt.ndt7_downloads                                                              
ndt.ndt5_uploads                                                                
ndt.ndt5_downloads                                                              
ndt.web100_uploads                                                              
ndt.web100_downloads                                                            
ndt.unified_uploads                                                             
ndt.unified_downloads
```

The split of upload/download doesn't need to happen until the last step. This still needs to remain manageable. This is getting more complex not less.

In hindsight, it's not clear that we want to publish the sidecar raw tables under `ndt.*` yet.
</blockquote></details>

The upload and download split is needed because the extended views are not the same.   Mixing upload and download makes some of the potential evolution nearly impossible because the sub expressions need to use different columns depending on direction.    We could union upload and download, but that encourages people to make the mistake of combining statistical data with very different distributions.

---


<!-- Sent from Reviewable.io -->

mattmathis commented:
___
*[views/intermediate_ndt/referenced-by/extended_ndt5_downloads.sql, line 124 at r1](https://reviewable.io/reviews/m-lab/etl-schema/91#-MKe_Fg9EcxFUZXuwWdI:-MKlw7Er1sevl1G3freJ:b4dbfda) ([raw file](https://github.com/m-lab/etl-schema/blob/23c70815e71c141d0a764181885e8b473f7f1606/views/intermediate_ndt/referenced-by/extended_ndt5_downloads.sql#L124)):*
<details><summary><i>Previously, stephen-soltesz (Stephen Soltesz) wrote…</i></summary><blockquote>

It is a mistake to introduce new record names here. The new fields should continue working without query changes after we retire the old fields. This is how the introduction of `date` and retirement of `test_date` worked.

For queries that currently use `Geo.continent_code`, we can either break now (and refer to the _<date> legacy table) or combine with the camel case so `Geo.ContinentCode` also works. We don't want users to have to change their queries twice if they adopt the new names now.
</blockquote></details>

Per our conversation I changed this to be _new_Geo, and added a comment to discourage people from accessing it directly (8 places total).   From a management standpoint it will be easier do delete all of Geo and rename _new_Geo to Geo rather than to do a similar transformation on a large number of individual columns.

One of the dangers of the method the we have been using (mixed old and new columns) is that people might overlook updating some columns, and then long after we and they think they have completed the migration, a MLab code cleanup causes their code to crash.  One of the cool features of having alternate structs rather than alternate columns is that I can support multiple versions, but effectively prohibit mixed version queries.   I am working on a seperate doc to describe this process, but we need to talk f2f.


<!-- Sent from Reviewable.io -->
